libnetwork/osl/route_linux.go

```diff
@@ -52,36 +52,33 @@ func (n *Namespace) setGatewayIPv6(gwv6 net.IP) {
 	n.mu.Unlock()
 }
 
-// SetGateway sets the default IPv4 gateway for the sandbox.
+// SetGateway sets the default IPv4 gateway for the sandbox. It is a no-op
+// if the given gateway is empty.
 func (n *Namespace) SetGateway(gw net.IP) error {
-	// Silently return if the gateway is empty
 	if len(gw) == 0 {
 		return nil
 	}
 
-	err := n.programGateway(gw, true)
-	if err == nil {
-		n.setGateway(gw)
+	if err := n.programGateway(gw, true); err != nil {
+		return err
 	}
-
-	return err
+	n.setGateway(gw)
```

Comment:
On one hand, this "just" locks `n.mu` and then set `n.gateway`; thus it serializes concurrent calls. OTOH, `n.programGateway` locks nothing and thus concurrent calls aren't serialized :thinking:

Reply:
Yeah, locking is a big mess in libnetwork (too many locks in one place, and no locks in places where it _actually_ matters). That's (see my other comment) also a reason I'm un-exporting the mutexes, which also makes it easier to find where locking happens.